back/app/services/survey_results_generator_service.rb

```diff
@@ -319,19 +402,211 @@ def normalise_domicile_keys(answer_groups, domicile_field)
+          input_type = question[:inputType]
+          page_number = question[:pageNumber]
+
+          # We only skip pages & questions from the next page onwards
+          if supports_page_logic? input_type
+            if page_number == next_page_number
+              next_page_number = nil
+              skip_question = false
+            elsif next_page_number
+              skip_question = true
+            end
+          end
+
+          # reduce the number of not answered if the question is skipped
+          if skip_question && question[:answers].present?
+            nil_answer = question[:answers].find { |a| a[:answer].nil? }
+            nil_answer[:count] -= 1 if nil_answer && nil_answer[:count] > 0
+          end
+
+          unless skip_question
+            # Only increment the number of times seen if we're not skipping the question/page
+            question[:questionViewedCount] += 1
+
+            # Calculate the next page number that will be seen
+            if supports_question_logic? input_type
+              answer_value = response[question[:key]]
+              values = answer_value.is_a?(Array) ? answer_value : [answer_value] # Convert all values to an array so all fields can be treated the same
+              highest_next_page_number_for_question = 0
+              values.each do |value|
+                logic_match = question.dig(:logic, :answer, value)
+                if logic_match && logic_match[:nextPageNumber] > highest_next_page_number_for_question
+                  # Only take the highest next page number from all options
+                  highest_next_page_number_for_question = logic_match[:nextPageNumber]
+                  next_page_number = highest_next_page_number_for_question
+                end
+              end
+            elsif supports_page_logic? input_type
+              logic_match = question[:logic][:nextPageNumber]
+              if logic_match
+                next_page_number = logic_match
+              end
+            end
+          end
+        end
       end
+    end
+
+    # Finalise the results
+    results.map do |question|
+      if survey_has_logic
+        # Update the total response count with the new figure
+        question[:totalResponseCount] = question[:questionViewedCount]
+
+        # Update the total pick count because we've reduced the 'not_answered' answer count
+        question[:totalPickCount] = question[:answers].pluck(:count).sum if question[:totalPickCount]
+      end
+    end
 
-      answer_titles[value][:title_multiloc].merge! labels
+    results
+  end
+
+  def cleanup_results(results)
+    # Remove the last page - only needed for logic calculations
+    results.pop if results.last[:inputType] == 'page' && results.last[:key] == 'survey_end'
+
+    # remove the temporary fields that are now not needed
+    results.map do |question|
+      question.delete(:questionViewedCount)
+      question.delete(:key)
+      question
     end
```

Comment:
Added this method to remove the last page, once we're done with it in the results